File: 달리기 경주.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;
map<string, int> arr;

vector<string> solution(vector<string> players, vector<string> callings) {
    for(int i=0;i<players.size();i++){
        string name=players[i];
        int rank=i;
        arr[name]=rank;
    }
    
    for(int i=0;i<callings.size();i++){
        
        int idx=arr[callings[i]];
        //kai를 접근할 때 밑의 코드로 인해 변경된 idx값으로 참조
        arr[players[idx-1]]++;
        //이전 친구는 랭크 더하고
        arr[players[idx]]--;
        //호명된 친구는 랭크 빼고
        swap(players[idx], players[idx-1]);
        
        
    }
    return players;
}

File: 2023-10-26/달리기경주.js
function solution(players, callings) {
  let runningPlayers = {};
  let playersNumber = {};

  players.forEach((e, i) => {
    const rank = i + 1;
    runningPlayers[e] = rank;
    playersNumber[rank] = e;
  });

  callings.forEach((e) => {
    const move = playersNumber[runningPlayers[e] - 1];
    playersNumber[runningPlayers[e]] = move;
    playersNumber[runningPlayers[move]] = e;
    runningPlayers[e] -= 1;
    runningPlayers[move] += 1;
  });
  console.log(playersNumber);
  return Object.values(playersNumber);
}

console.log(
  solution(
    ['mumu', 'soe', 'poe', 'kai', 'mine'],
    ['kai', 'kai', 'mine', 'mine']
  )
);
